enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,344 @@
+- Deploying the StatefulSet otherwise works just like in the base case.
+
+#### Certificate expiration, manual secret regeneration, CA rotations
+
+Works substantially the same as for Service certificates generated by the Service CA
+(notably currently requiring the admins to manually trigger a restart of affected pods
+after certificates are regenerated).
+
+### Implementation Details/Notes/Constraints
+
+#### The Pod Creation vs. Secret Creation Race
+
+As described above,
+creating a StatefulSet triggers both pods and the secret to be created concurrently,
+and the application must be designed to support that by triggering the pod to be re-created.
+
+To avoid the race in predictable situations,
+the controller creates some _extra_ certificates on top of the StatefulSet.spec.replicas request,
+so that a future small scale-up can succeed at the first try.
+
+The implementation currently hard-codes creating 5 extra certificates,
+or 30% more than spec.replicas, whichever is larger.
+
+If the StatefulSet is scaled down, we don’t immediately delete the extra certificates;
+that would just be extra traffic for no benefit.
+The secret is trimmed to a smaller size only when re-created
+(on certificate expiry or a resync, e.g. if the user manually deletes or corrupts the secret).
+
+#### Annotations
+
+StatefulSets now use the same annotations as Services (`serving-sert-secret-name` as input, `serving-cert-{created-by,generation-error,generation-error-num}` as output.
+
+Secrets can now have `originating-StatefulSet-{name,uid}` annotations
+to link back to the StatefulSet;
+this works just like the previous `originating-service-*` annotations.
+
+##### Alpha Annotations Not Used
+
+The StatefulSet controllers only use `service.beta.openshift.io` annotations,
+not the `alpha` annotation which are supported for services.
+This is new code, so supporting an older version seems unnecessary.
+(Or should this feature start out as `alpha`?)
+
+#### The OpenShift “Service UID” Certificate Extension Is Not Used
+The service-ca certificates created for services
+include an OpenShift-specific “Service UID” extension,
+per https://github.com/openshift/origin/pull/12413 to identify intra-cluster ElasticSearch traffic.
+
+That extension is specifically documented to contain a service UID,
+so it can’t be directly used for StatefulSets,
+and it’s not obvious that looking up the UID of a service
+referenced by StatefulSet.spec.serviceName makes much sense —
+that StatefulSet→Service reference is UID-agnostic,
+and it’s a different set of certificates
+that was not anticipated by the current consumers of the “service UID” certificate extension.
+
+If necessary, we can allocate and define a new “StatefulSet UID” extension,
+or look up a Service UID for StatefulSet.spec.serviceName,
+or maybe even store a StatefulSet UID in the “Service UID”  extension, in the future;
+it seems unnecessary for a first version of this feature.
+
+#### Secret Size Grows with Replicas
```

Comment:
So this scale to 100'ish nodes before reaching 2 MB object size. I don't think there are many instances of stateful services of that size. And in that case you have probably other problems than using an in-cluster mechanism like service-ca.

Reply:
Right now, the controller code gives up after (hard-coded) 10 failures creating a secret, and I think tries again (another 10 tries) on a complete re-sync that the library-go controller infrastructure causes every 20 minutes.

I’ll update this to use the `serving-cert-generation-error-num` annotation, so that repeated failures creating a secret cause the controller to give up permanently.